src/hera/shared/_pydantic.py

```diff
@@ -1,8 +1,39 @@
 """Module that holds the underlying base Pydantic models for Hera objects."""
-from pydantic import BaseModel as PyBaseModel
 
-
-class BaseModel(PyBaseModel):
+# The pydantic v1 interface is used for both pydantic v1 and v2 in order to support
+# users across both versions.
+try:
+    from pydantic.v1 import (
+        BaseModel as PydanticBaseModel,
+        Field,
+        ValidationError,
+        root_validator,
+        validate_arguments,
+        validator,
+    )
```

Comment:
Might be worth talking about how we, as maintainers of Hera, should be using Pydantic v2 - eventually we'll need to use v2 classes and properly migrate. Is there any way we can do that gradually/make it easier than an eventual big bang major version upgrade? cc @flaviuvadan

Reply:
Perhaps we decide on which breaking changes should go into v6, accumulate a few of those, this one included, and then migrate to V2? 🤷🏻

Reply:
Might be a good time to also move Hera to Argo 3.5